### DISCUSSION TITLE
flash and render

### DISCUSSION THREAD

**Discussion by fluca1978** [mojolicious/mojo#2161] [Q&A] [answered]
I'm probably misunderstanding `flash` and its usage with `render`. In my controller method I've something like:

```
sub logout( $self ) {
    $self->flash( message => "Logout done",
		  message_class => 'alert_success' );

    $self->render( template => q{welcome/welcome},
		   format  => q/html/, );
}
```

and my layout I've:

```
    <% if ( my $message = flash 'message' ) { %>
    <div class="alert <%= flash 'message_class' %> alert-dismissible fade show" role="alert">
      <strong>
	<%= $message %>
      </strong>
      <button type="button" class="close" data-dismiss="alert" aria-label="Close">
	<span aria-hidden="true">&times;</span>
      </button>
    </div>
    <% } %>

```

but the flash message is displayed only if I reload the same page aftet it has been rendered the first time. In the `Mojolicious::Plugin::DefaultHelpers` documentation the example uses a `redirect_to` after setting the flash.and it clearly works, but what if I want to use the flash mechanism for `render`? Do I need to use the stash instead?

**Comment by kraih** [comment, marked as answer]
The `flash` feature is specifically meant for use with redirects. Without a redirect you have no need for it. Don't get it confused with the `stash`.

**Reply by guest20** [reply]
The old flash and stash trick, that's the second time I've used that this month.
  ~~ maxwell smart, probably
